src/hotspot/os/posix/perfMemory_posix.cpp

```diff
@@ -727,48 +715,97 @@ static void cleanup_sharedmem_resources(const char* dirname) {
-    // then remove the stale file resources.
+    // We now have a file name that converts to a valid integer
+    // that could represent a process id.
+    //
+    // On Linux, we first try to flock() on the file to check
+    // for JVM processes in other containers that share the same
+    // /tmp directory as the current process. If it's already
+    // locked by another process, then obviously it's not stale.
+    //
+    // Otherwise, if this process id matches the current process id or
+    // the process is not running, we treat it as a stale file and remove it.
     //
-    // process liveness is detected by sending signal number 0 to
-    // the process id (see kill(2)). if kill determines that the
-    // process does not exist, then the file resources are removed.
-    // if kill determines that that we don't have permission to
-    // signal the process, then the file resources are assumed to
-    // be stale and are removed because the resources for such a
+    // Process liveness is detected by sending signal number 0 to
+    // the process id (see kill(2)). If kill determines that the
+    // process does not exist, then the file is removed.
+    // If kill determines that that we don't have permission to
+    // signal the process, then the file is assumed to
+    // be stale and is removed because the files for such a
     // process should be in a different user specific directory.
     //
-    if ((pid == os::current_process_id()) ||
-        (kill(pid, 0) == OS_ERR && (errno == ESRCH || errno == EPERM))) {
-        unlink(entry->d_name);
+
+    bool is_locked_by_another_process = false;
+ #if defined(LINUX)
+    // The locking protocol works only with JVMs that have the JDK-8286030 fix.
+    // If you are sharing the /tmp difrectory among different containers, do not
+    // use older JVMs that don't have this fix.
+    int fd;
+    RESTARTABLE(os::open(filename, O_RDONLY, 0), fd);
+    if (fd == OS_ERR) {
+      // Something wrong happened. Ignore the error and don't try to remove the
+      // file.
+      continue;
     }
+    int n;
+    RESTARTABLE(::flock(fd, LOCK_EX|LOCK_NB), n);
+    if (n != 0) {
+      if (errno == EWOULDBLOCK) {
+        // Another process holds the exclusive lock on this file.
+        is_locked_by_another_process = true;
+      } else {
+        // Something wrong happened. Ignore the error and don't try to remove the
+        // file.
+        ::close(fd);
+        continue;
+      }
+    } else {
+      // I have successfully locked the file. No other process will try to write to
+      // it, so I can safely remove it below.
+    }
+    log_info(perf, memops)("is_locked_by_another_process %s/%s = %s", dirname, filename,
+                            is_locked_by_another_process ? "true" : "false");
+#endif
+
+    if (!is_locked_by_another_process) {
+      if ((pid == os::current_process_id()) ||
+          (kill(pid, 0) == OS_ERR && (errno == ESRCH || errno == EPERM))) {
```

Comment:
Thinking about this, I now was confused. AFAICS this code is only ever called from `mmap_create_shared` with the directory containing our own pid. I do not see this called for PIDs from other processes. Why do we handle that case? Or am I overlooking something?

Reply:
The current JVM process only cleans up stale files in the current user's own directory. I.e., /tmp/hsperfdata_bob/. However,  when we see a file like /tmp/hsperfdata_bob/1234, it's possible that

- 1234 was a JVM process created by bob, but this process died without cleaning up
- A new process 1234 is now running, but it belongs to a different user, john

So if `kill(1234, 0)` fails, the JVM concludes that, "/tmp/hsperfdata_bob/1234 must belong to a terminated JVM process owned by bob".

I think this is the intention of the code, in spite of what the comment says above the `kill` call.

Reply:
I simplified the code and restructured the comments to be interspersed with the code. Hopefully that would make the logic a little easier to understand.

Despite the large diff in cleanup_sharedmem_files(), the only actual change is to `flock` in the middle of the loop, and avoid deleting the file if the `flock` fails.

I removed this comment

```
// ... are removed because the resources for such a
// process should be in a different user specific directory.
```

and replaced it with something more comprehensive which hopefully makes more sense:

```
// This directory should be used only by JVM processes owned by the
// current user to store PerfMemory files. Any other files found
// in this directory may be removed.
```